lib/matplotlib/backends/qt_compat.py

```diff
@@ -1,220 +1,159 @@
+        from PyQt4 import QtCore, QtGui
+        __version__ = QtCore.PYQT_VERSION_STR
+        # PyQt 4.6 introduced getSaveFileNameAndFilter:
+        # https://riverbankcomputing.com/news/pyqt-46
+        if __version__ < LooseVersion("4.6"):
+            raise ImportError("PyQt<4.6 is not supported")
+        QtCore.Signal = QtCore.pyqtSignal
         QtCore.Slot = QtCore.pyqtSlot
-    except AttributeError:
-        # Not a perfect match but works in simple cases
-        QtCore.Slot = QtCore.pyqtSignature
-
-    QtCore.Property = QtCore.pyqtProperty
-    __version__ = QtCore.PYQT_VERSION_STR
+        QtCore.Property = QtCore.pyqtProperty
+        _getSaveFileName = QtGui.QFileDialog.getSaveFileNameAndFilter
 
-# Apply shim to Qt4 APIs to make them look like Qt5
-if QT_API in (QT_API_PYQT, QT_API_PYQTv2, QT_API_PYSIDE):
-    '''Import all used QtGui objects into QtWidgets
-
-    Here I've opted to simple copy QtGui into QtWidgets as that
-    achieves the same result as copying over the objects, and will
-    continue to work if other objects are used.
-
-    '''
+    if QT_API == QT_API_PYQTv2:
+        _setup_pyqt4_internal(api=2)
+    elif QT_API == QT_API_PYSIDE:
+        from PySide import QtCore, QtGui, __version__, __version_info__
+        # PySide 1.0.3 fixed the following:
+        # https://srinikom.github.io/pyside-bz-archive/809.html
+        if __version_info__ < (1, 0, 3):
+            raise ImportError("PySide<1.0.3 is not supported")
+        _getSaveFileName = QtGui.QFileDialog.getSaveFileName
+    elif QT_API == QT_API_PYQT:
+        _setup_pyqt4_internal(api=1)
+    else:
+        raise ValueError("Unexpected value for the 'backend.qt4' rcparam")
     QtWidgets = QtGui
 
+    def is_pyqt5():
+        return False
+
+
+if QT_API in [QT_API_PYQT5, QT_API_PYSIDE2]:
+    _setup_pyqt5()
+elif QT_API in [QT_API_PYQTv2, QT_API_PYSIDE, QT_API_PYQT]:
+    _setup_pyqt4()
+elif QT_API is None:
+    if rcParams["backend"] == "Qt4Agg":
+        _candidates = [(_setup_pyqt4, QT_API_PYQTv2),
+                       (_setup_pyqt4, QT_API_PYSIDE),
+                       (_setup_pyqt4, QT_API_PYQT),
+                       (_setup_pyqt5, QT_API_PYQT5),
+                       (_setup_pyqt5, QT_API_PYSIDE2)]
+    else:
+        _candidates = [(_setup_pyqt5, QT_API_PYQT5),
+                       (_setup_pyqt5, QT_API_PYSIDE2),
+                       (_setup_pyqt4, QT_API_PYQTv2),
+                       (_setup_pyqt4, QT_API_PYSIDE),
+                       (_setup_pyqt4, QT_API_PYQT)]
+    for _setup, QT_API in _candidates:
+        try:
+            _setup()
+        except ImportError:
+            continue
+        break
+    else:
+        raise ImportError("Failed to import any qt binding")
+else:
+    raise AssertionError  # We should not get there.
+
 
-def is_pyqt5():
-    return QT_API == QT_API_PYQT5
+# These globals are only defined for backcompatibilty purposes.
+ETS = dict(pyqt=(QT_API_PYQTv2, 4), pyside=(QT_API_PYSIDE, 4),
```

Comment:
Copy the comments from the original ETS definition?

Reply:
The dict is basically irrelevant now and only left for backcompat; there's a similar comment at the top ("Otherwise check the QT_API environment variable etc.").